WalletWasabi/Blockchain/Transactions/SmartTransaction.cs

```diff
@@ -32,23 +35,52 @@ public SmartTransaction(Transaction transaction, Height height, uint256? blockHa
 		FirstSeen = firstSeen == default ? DateTimeOffset.UtcNow : firstSeen;
 
 		IsReplacement = isReplacement;
-		WalletInputs = new HashSet<SmartCoin>(Transaction.Inputs.Count);
-		WalletOutputs = new HashSet<SmartCoin>(Transaction.Outputs.Count);
+		_walletInputs = new HashSet<SmartCoin>(Transaction.Inputs.Count);
+		_walletOutputs = new HashSet<SmartCoin>(Transaction.Outputs.Count);
 	}
 
 	#endregion Constructors
 
 	#region Members
 
-	/// <summary>
-	/// Coins those are on the input side of the tx and belong to ANY loaded wallet. Later if more wallets are loaded this list can increase.
-	/// </summary>
-	public HashSet<SmartCoin> WalletInputs { get; }
+	public IReadOnlyCollection<SmartCoin> WalletInputs
+	{
+		get
+		{
+			return _walletInputs;
+		}
+	}
 
-	/// <summary>
-	/// Coins those are on the output side of the tx and belong to ANY loaded wallet. Later if more wallets are loaded this list can increase.
-	/// </summary>
-	public HashSet<SmartCoin> WalletOutputs { get; }
+	public void AddWalletInput(SmartCoin input)
+	{
+		_walletInputs.Add(input);
+		_foreignInputs = null;
+		_walletVirtualInputs = null;
```

Comment:
Why null out everything else every time a wallet input is added? It seems like you add something then null out unrelated things. (Same question applies for `AddWalletOutput` and `RemoveWalletOutput`)

Reply:
When you add a wallet input you should remove this input from the list of foreign inputs and update the list of wallet virtual inputs.

Instead of doing this operation every time a wallet input is added
  * I set `_foreignInputs` and `_walletVirtualInputs` to `null` which indicates these variables are no longer valid and
  * recalculate the variables only when they are needed.

Reply:
That makes perfect sense, thank you! Can you add comments so the code readers don't get confused?

Reply:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/7932/commits/d0bbc26fc79e37117efc977f3393421190b1cb4a.